src/PIL/PdfImagePlugin.py

```diff
@@ -151,38 +101,70 @@ def write(self, value):
                 # Image format does not have n_frames. It is a single frame image
                 pass
         numberOfPages += im_numberOfPages
-    pages = [str(pageNumber*3+4)+" 0 R"
-             for pageNumber in range(0, numberOfPages)]
+        for i in range(im_numberOfPages):
+            image_refs.append(existing_pdf.next_object_id(0))
+            page_refs.append(existing_pdf.next_object_id(0))
+            contents_refs.append(existing_pdf.next_object_id(0))
+            existing_pdf.pages.append(page_refs[-1])
 
-    xref.append(fp.tell())
-    _obj(
-        fp, 2,
-        Type="/Pages",
-        Count=len(pages),
-        Kids="["+"\n".join(pages)+"]")
-    _endobj(fp)
+    #
+    # catalog and list of pages
+    existing_pdf.write_catalog()
 
     pageNumber = 0
     for imSequence in ims:
         for im in ImageSequence.Iterator(imSequence):
+            # FIXME: Should replace ASCIIHexDecode with RunLengthDecode (packbits)
+            # or LZWDecode (tiff/lzw compression).  Note that PDF 1.2 also supports
```

Comment:
Just a note that I don't trust any of the pillow internal packbits or LZW compression methods to be correct, as they appear to lead to corrupt images in some of the tiff tests. We've been moving off of them in favor of libtiff. 

Flatedecode should be ok, as that's a a library function already.

Reply:
Understood, thanks for the heads up. This is just moving around code that was already there, though, so I don't feel like doing anything about it in this pull request.

Reply:
Fair enough.